lisa/tpp_model.py

```diff
@@ -0,0 +1,634 @@
+
+    def trace_sim_activation_complete(self, pe, task, start, end):
+        self.buffers['sim_activation_complete'].append((self.unique_time_stamp(), pe.cpuid, pe.task.pid, pe.task.name, pe.cpuid, task.pid, start, end))
+
+
+class TaskState():
+    TASK_RUNNING = 0x0000
+    TASK_INTERRUPTIBLE = 0x0001
+    TASK_UNINTERRUPTIBLE = 0x002
+
+class SimulationConf(SimpleMultiSrcConf):
+    """
+    YAML configuration helper.
+    """
+    CPUCapacities = TypedDict[int,int]
+    FreqList = SortedTypedList[int]
+    CPUIdList = SortedTypedList[int]
+
+    STRUCTURE = TopLevelKeyDesc('tpp-conf', 'Simulation configuration', (
+        KeyDesc('name', 'Simulation instance name', [str]),
+        LevelKeyDesc('cpu', 'PE properties', (
+            KeyDesc('capacities', 'Compute capacities of CPUs', [CPUCapacities]),
+            KeyDesc('freqs', 'CPU performance levels', [TypedDict[int,FreqList]]),
+            KeyDesc('perf-domains', 'CPU performance domain', [TypedList[CPUIdList]]),
+        )),
+        LevelKeyDesc('workload', 'workload properties', (
+            KeyDesc('tasks', 'tasks (busy_time [s], period [s], offset [s])', [TypedDict[int,TypedList[float]]]),
+        ))
+    ))
+
+class Simulation(Configurable, Loggable):
+    """
+    Sets up and configures Simulation.
+    """
+    CONF_CLASS = SimulationConf
+    INIT_KWARGS_KEY_MAP = {
+        'cpu_capacities': ['cpu','capacities'],
+        'cpu_freqs': ['cpu','freqs'],
+        'cpu_perf_domains': ['cpu','perf-domains'],
+        'task_defs': ['workload','tasks'],
+    }
+
+    def __init__(self, name='<noname>', cpu_capacities=None, cpu_freqs=None, cpu_perf_domains=None, task_defs=None):
+        super().__init__()
+        self.name = name
+        self.cpu_capacities = cpu_capacities
+        self.cpu_freqs = cpu_freqs
+        self.cpu_perf_domains_conf = cpu_perf_domains
+
+        self.pes = {}
+        self.env = simpy.Environment()
+        self.trace = Trace(self.env)
+        self.tasks = []
+
+        self.scaling_govs = []
+        self.cpu_perf_doms = []
+
+        self._create_system()
+        self._setup_workload(task_defs)
+
+    @classmethod
+    def from_conf(cls, conf, **kwargs):
+        kwargs = cls.conf_to_init_kwargs(conf)
+        cls.check_init_param(**kwargs)
+        print(conf.pretty_format())
+
+        return cls(**kwargs)
+
+    def _create_system(self):
+        for pe in self.cpu_capacities.items():
+            cpuid, capacity = pe
+            self.pes[cpuid] = PE(self.env, cpuid, self.trace, capacity=capacity, dvfs_capacities=self.cpu_freqs[cpuid])
+
+        self.scheduler = Scheduler(self.env, self.pes.values(), self.trace)
+
+        for perf_dom in self.cpu_perf_domains_conf:
+            pes = {}
+            for cpuid in perf_dom:
+                pes[cpuid] = self.pes[cpuid]
+
+            self.scaling_govs.append(Capacity_Scaling_gov(self.env, pes))
```

Comment:
Found another solution: parametrizing the whole module so that the user can inject their own custom class into it. The idea is very close to ML languages family modules: the module is coded against a set of abstract APIs, and it's up to the user to instanciate the module to provide concrete implementations. This has very little impact on the coding style of the module itself, and allows full flexibility (we can even inherit from user-provided classes if needed).

The main drawback of this system is that serialization (pickle or YAML) of instances of classes defined in such module will probably not work, because a given full class name (e.g. `lisa.conf.MultiSrcConf`) can map to an arbitrary number of different types (one per instance of the module). The actual classes the deserializer would attempt to use when looking up that name would be impossible to instantiate, as they would make use of abstract base classes. If that is really necessary, we could probably alleviate that problem for "canned" instances of the module, that would be directly accessible without user instantiation. 

I have implemented it at: https://github.com/ARM-software/lisa/pull/1722